### PR TITLE
tkt-60768: Break ioc_json into more classes

### PR DESCRIPTION
**Pull request by skarekrow**
Possibly more can be moved into the new classes, but for now these were needed.

- Fix bad release prop setting
- Reduce all usage of getting pool and iocroot many times

Closes iocage#700
FreeNAS Ticket: #60768

### CODE REVIEW

File: iocage_lib/iocage.py [hunk 1]
```diff
@@ -90,8 +90,7 @@ def get_iocroot(self):
         Return:
                 string: with the iocroot name.
         """
-
-        return ioc_json.IOCJson(self.pool).json_get_value("iocroot")
+        return ioc_json.IOCJson().json_get_value("iocroot")
 
 
 class IOCage(object):
```
Comment (sonicaj) [inline iocage_lib/iocage.py]: Should we declare `self.zfs` from the zfs class we have created ?

Comment (sonicaj) [reply iocage_lib/iocage.py]: Actually, I think that would make more sense in another PR where we can consolidate zfs related functionality. Thoughts ?

File: iocage_lib/ioc_json.py [hunk 2]
```diff
@@ -44,19 +44,672 @@
...
+                            "INFO",
+                            "message":
+                            f"Setting up zpool [{zpool}] for"
+                            " iocage usage\n If you wish to change"
+                            " please use \"iocage activate\""
+                        },
+                        _callback=self.callback,
+                        silent=self.silent)
+
+                    self.zfs_set_property(zpool, "org.freebsd.ioc:active",
```
[the remark this answers is not in the document]
Comment (skarekrow) [reply iocage_lib/ioc_json.py]: We could but that's unrelated to this ;) The idea is with https://github.com/freenas/iocage/pull/125/files/2c48392ca7c58689dfcc105098239e2ac5ab02cf#diff-f0920b7a44ac8a00eec7012b7a345093R99 that they will be active/valid pools.

File: iocage_lib/ioc_json.py [hunk 3]
```diff
@@ -639,8 +1097,7 @@ def json_get_value(self, prop, default=False):
                 try:
                     return conf[prop]
                 except KeyError:
-                    default_props = self.json_check_default_config()
-                    return default_props[prop]
+                    return self.default_config[prop]
 
             if write:
                 self.json_write(conf)
```
[the remark this answers is not in the document]
Comment (skarekrow) [reply iocage_lib/ioc_json.py]: If the defaults changed then they won't be written without it. There's also the problem of overwriting corrupt user defaults, which we don't want to do.

File: iocage_lib/ioc_json.py [hunk 4]
```diff
@@ -778,388 +1234,117 @@ def json_set_value(self, prop, _import=False, default=False):
...
-                            "EXCEPTION",
+                            "level": "EXCEPTION",
                             "message":
-                            "A renaming operation is required to continue.\n"
-                            f"Please run iocage -f {sys.argv[-1]} as root."
+                            f"{uuid} is running.\nPlease stop it first!"
                         },
                         _callback=self.callback,
                         silent=self.silent)
+            elif key == "dhcp":
```
[the remark this answers is not in the document]
Comment (skarekrow) [reply iocage_lib/ioc_json.py]: The previous if is for the jail_zfs_* props. So putting dhcp there wouldn't happen *unless* they were also specified.

File: iocage_lib/ioc_json.py [hunk 5]
```diff
@@ -778,388 +1234,117 @@ def json_set_value(self, prop, _import=False, default=False):
...
+            if status:
+                if key in single_period:
+                    key = key.replace("_", ".", 1)
+                else:
+                    key = key.replace("_", ".")
 
-                if rtrn:
-                    # They want to stop the jail, not attempt to migrate before
+                if key in jail_params:
+                    if full_conf["vnet"] == "on" and key == "ip4.addr" or \
```
[the remark this answers is not in the document]
Comment (skarekrow) [reply iocage_lib/ioc_json.py]: We could, but style wise I don't like brackets ;)

[the remark this answers is not in the document]
Comment (sonicaj) [reply iocage_lib/ioc_json.py]: Just to clarify, logically we want to return from this function when `vnet` is `on` and `key` is either `ip4.addr` or `ip6.addr` ?

[the remark this answers is not in the document]
Comment (skarekrow) [reply iocage_lib/ioc_json.py]: Yep!

[the remark this answers is not in the document]
Comment (sonicaj) [reply iocage_lib/ioc_json.py]: I am not sure if that is intended behaviour ?

File: iocage_lib/ioc_json.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (skarekrow) [reply iocage_lib/ioc_json.py]: haha good point!

File: iocage_lib/ioc_json.py [hunk 6]
```diff
@@ -44,19 +44,672 @@
...
+                'freebsd-version'
+        except FileNotFoundError:
+            if template == 'yes':
+                freebsd_version = f'{self.iocroot}/templates/' \
+                    f"{conf['host_hostuuid']}/root/bin/freebsd-version"
+            else:
+                temp_uuid = self.location.rsplit('/', 1)[-1]
+                freebsd_version = f'{self.iocroot}/jails/{temp_uuid}' \
+                    '/root/bin/freebsd-version'
+        except KeyError:
```
Comment (sonicaj) [inline iocage_lib/ioc_json.py]: I am not sure where a `KeyError` could be raised here ?

Comment (sonicaj) [reply iocage_lib/ioc_json.py]: If we are looking at key `host_hostuuid` of `conf`, I think the `KeyError` catch clause will have to be moved under `FileNotFoundError` as the current one would not cover it

File: iocage_lib/ioc_json.py [hunk 7]
```diff
@@ -44,19 +44,672 @@
...
+                _callback=self.callback,
+                silent=self.silent)
+
+        release = release.rsplit('-p', 1)[0]
+        cloned_release = conf.get('cloned_release', 'LEGACY_JAIL')
+
+        try:
+            freebsd_version = f'{self.iocroot}/releases/{release}/root/bin/' \
+                'freebsd-version'
+        except FileNotFoundError:
```
Comment (sonicaj) [inline iocage_lib/ioc_json.py]: This exception would never be raised ?

File: iocage_lib/ioc_json.py [hunk 8]
```diff
@@ -44,19 +44,672 @@
...
+
+                    return conf
+
+        except KeyError:
+            # New jail creation
+            pass
+
+        try:
+            if not renamed:
+                self.json_write(conf)
```
Comment (sonicaj) [inline iocage_lib/ioc_json.py]: Not sure if `json_write` can raise `FileNotFoundError` ? I think it will write to it if it doesn't exist

**Pull request merged**